File: bloom_filter.ml
(* bloom_filter.ml — Probabilistic Set Membership *)
(* A Bloom filter uses k hash functions over an m-bit array to test     *)
(* set membership with tunable false-positive rate. No false negatives. *)
(* Supports: create, add, mem, count, union, false_positive_rate,      *)
(* saturation, of_list, mem_all, mem_any, clear, copy, serialize.      *)
(* Immutable — all operations return new filters.                      *)

module BloomFilter = struct

  type t = {
    bits : bytes;     (* bit array stored as bytes *)
    m    : int;       (* total bits *)
    k    : int;       (* number of hash functions *)
    n    : int;       (* number of elements added *)
  }

  (* ---- Bit manipulation helpers ---- *)

  let get_bit bits i =
    let byte_idx = i / 8 in
    let bit_idx = i mod 8 in
    Char.code (Bytes.get bits byte_idx) land (1 lsl bit_idx) <> 0

  let set_bit bits i =
    let byte_idx = i / 8 in
    let bit_idx = i mod 8 in
    let old = Char.code (Bytes.get bits byte_idx) in
    Bytes.set bits byte_idx (Char.chr (old lor (1 lsl bit_idx)))

  (* ---- Hash functions ---- *)
  (* Double hashing: h_i(x) = (h1(x) + i * h2(x)) mod m *)
  (* Uses Hashtbl.hash with salt for h1, h2 *)

  let hash1 x = Hashtbl.hash x
  let hash2 x = Hashtbl.hash (x, 0x9e3779b9)

  let hash_i m h1 h2 i =
    ((h1 + i * h2) mod m + m) mod m

  (* ---- Core API ---- *)

  (** Create an empty Bloom filter.
      [m] = number of bits, [k] = number of hash functions. *)
  let create ?(m = 1024) ?(k = 7) () =
    let m = max 8 m in
    let k = max 1 k in
    let byte_count = (m + 7) / 8 in
    { bits = Bytes.make byte_count '\000'; m; k; n = 0 }

  (** Create a Bloom filter sized for expected element count and
      desired false-positive probability. *)
  let create_optimal ~expected_elements ~fp_rate =
    let n_f = float_of_int (max 1 expected_elements) in
    let p = max 0.0001 (min 0.5 fp_rate) in
    let m_f = -.(n_f *. log p) /. (log 2.0 ** 2.0) in
    let m = max 8 (int_of_float (ceil m_f)) in
    let k_f = (float_of_int m /. n_f) *. log 2.0 in
    let k = max 1 (int_of_float (Float.round k_f)) in
    create ~m ~k ()

  (** Add an element to the filter. Returns a new filter. *)
  let add x bf =
    let new_bits = Bytes.copy bf.bits in
    let h1 = hash1 x in
    let h2 = hash2 x in
    for i = 0 to bf.k - 1 do
      let idx = hash_i bf.m h1 h2 i in
      set_bit new_bits idx
    done;
    { bf with bits = new_bits; n = bf.n + 1 }

  (** Test if an element might be in the set.
      Returns true if possibly present, false if definitely absent. *)
  let mem x bf =
    let h1 = hash1 x in
    let h2 = hash2 x in
    let rec check i =
      if i >= bf.k then true
      else
        let idx = hash_i bf.m h1 h2 i in
        if get_bit bf.bits idx then check (i + 1)
        else false
    in
    check 0

  (** Number of elements added. *)
  let count bf = bf.n

  (** Number of bits in the filter. *)
  let size bf = bf.m

  (** Number of hash functions. *)
  let num_hashes bf = bf.k

  (** Count of set bits. *)
  let popcount bf =
    let c = ref 0 in
    for i = 0 to bf.m - 1 do
      if get_bit bf.bits i then incr c
    done;
    !c

  (** Saturation: fraction of bits that are set (0.0 to 1.0). *)
  let saturation bf =
    float_of_int (popcount bf) /. float_of_int bf.m

  (** Theoretical false-positive probability given current fill. *)
  let false_positive_rate bf =
    if bf.n = 0 then 0.0
    else
      let m_f = float_of_int bf.m in
      let k_f = float_of_int bf.k in
      let n_f = float_of_int bf.n in
      (1.0 -. exp (-. k_f *. n_f /. m_f)) ** k_f

  (** Is the filter empty (no elements added)? *)
  let is_empty bf = bf.n = 0

  (** Union of two filters (must have same m and k). *)
  let union bf1 bf2 =
    if bf1.m <> bf2.m || bf1.k <> bf2.k then
      invalid_arg "BloomFilter.union: incompatible filters"
    else begin
      let new_bits = Bytes.copy bf1.bits in
      let byte_count = (bf1.m + 7) / 8 in
      for i = 0 to byte_count - 1 do
        let b1 = Char.code (Bytes.get bf1.bits i) in
        let b2 = Char.code (Bytes.get bf2.bits i) in
        Bytes.set new_bits i (Char.chr (b1 lor b2))
      done;
      { bits = new_bits; m = bf1.m; k = bf1.k; n = bf1.n + bf2.n }
    end

  (** Create a filter from a list of elements. *)
  let of_list ?(m = 1024) ?(k = 7) lst =
    List.fold_left (fun bf x -> add x bf) (create ~m ~k ()) lst

  (** Test if all elements in a list might be present. *)
  let mem_all lst bf =
    List.for_all (fun x -> mem x bf) lst

  (** Test if any element in a list might be present. *)
  let mem_any lst bf =
    List.exists (fun x -> mem x bf) lst

  (** Create a cleared copy (same m and k, no elements). *)
  let clear bf = create ~m:bf.m ~k:bf.k ()

  (** Create a deep copy. *)
  let copy bf = { bf with bits = Bytes.copy bf.bits }

  (** Serialize to string (for debugging). *)
  let to_string bf =
    Printf.sprintf "BloomFilter(m=%d, k=%d, n=%d, saturation=%.2f%%)"
      bf.m bf.k bf.n (saturation bf *. 100.0)
end

File: Try/scripts/bloomFilter.js
/**
 * bloom_filter.js — JavaScript port of bloom_filter.ml
 *
 * Faithful translation of the OCaml Bloom filter implementation.
 * Supports: create, createOptimal, add, mem, count, size, numHashes,
 * popcount, saturation, falsePositiveRate, isEmpty, union, ofList,
 * memAll, memAny, clear, copy, toString.
 *
 * Immutable — all mutation-like operations return new filter objects.
 */
'use strict';

/**
 * Simple string hash (djb2 variant).
 * @param {*} x - value to hash
 * @returns {number}
 */
function hash1(x) {
  const s = String(x);
  let h = 5381;
  for (let i = 0; i < s.length; i++) {
    h = ((h << 5) + h + s.charCodeAt(i)) | 0;
  }
  return h >>> 0;
}

/**
 * Second hash function (FNV-1a variant).
 * @param {*} x - value to hash
 * @returns {number}
 */
function hash2(x) {
  const s = String(x);
  let h = 2166136261;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h >>> 0;
}

/**
 * Double hashing: h_i(x) = (h1(x) + i * h2(x)) mod m
 */
function hashI(m, h1, h2, i) {
  return ((h1 + i * h2) % m + m) % m;
}

/**
 * Get bit at position i in Uint8Array.
 */
function getBit(bits, i) {
  const byteIdx = Math.floor(i / 8);
  const bitIdx = i % 8;
  return (bits[byteIdx] & (1 << bitIdx)) !== 0;
}

/**
 * Set bit at position i in Uint8Array (mutates the array).
 */
function setBit(bits, i) {
  const byteIdx = Math.floor(i / 8);
  const bitIdx = i % 8;
  bits[byteIdx] = bits[byteIdx] | (1 << bitIdx);
}

/**
 * Create an empty Bloom filter.
 * @param {object} [opts]
 * @param {number} [opts.m=1024] - number of bits
 * @param {number} [opts.k=7]   - number of hash functions
 * @returns {object} Bloom filter
 */
function create({ m = 1024, k = 7 } = {}) {
  m = Math.max(8, m);
  k = Math.max(1, k);
  const byteCount = Math.ceil(m / 8);
  return Object.freeze({
    bits: new Uint8Array(byteCount),
    m,
    k,
    n: 0,
  });
}

/**
 * Create a Bloom filter sized for expected element count and
 * desired false-positive probability.
 * @param {number} expectedElements
 * @param {number} fpRate - desired false positive rate (0 < fpRate < 1)
 * @returns {object} Bloom filter
 */
function createOptimal(expectedElements, fpRate) {
  const nf = Math.max(1, expectedElements);
  const p = Math.max(0.0001, Math.min(0.5, fpRate));
  const mf = -(nf * Math.log(p)) / (Math.log(2) ** 2);
  const m = Math.max(8, Math.ceil(mf));
  const kf = (m / nf) * Math.log(2);
  const k = Math.max(1, Math.round(kf));
  return create({ m, k });
}

/**
 * Add an element to the filter. Returns a new filter.
 */
function add(x, bf) {
  const newBits = new Uint8Array(bf.bits);
  const h1v = hash1(x);
  const h2v = hash2(x);
  for (let i = 0; i < bf.k; i++) {
    const idx = hashI(bf.m, h1v, h2v, i);
    setBit(newBits, idx);
  }
  return Object.freeze({ bits: newBits, m: bf.m, k: bf.k, n: bf.n + 1 });
}

/**
 * Test if an element might be in the set.
 * Returns true if possibly present, false if definitely absent.
 */
function mem(x, bf) {
  const h1v = hash1(x);
  const h2v = hash2(x);
  for (let i = 0; i < bf.k; i++) {
    const idx = hashI(bf.m, h1v, h2v, i);
    if (!getBit(bf.bits, idx)) return false;
  }
  return true;
}

function count(bf) { return bf.n; }
function size(bf) { return bf.m; }
function numHashes(bf) { return bf.k; }

/**
 * Count of set bits (Kernighan's trick).
 */
function popcount(bf) {
  const byteCount = Math.ceil(bf.m / 8);
  let c = 0;
  for (let i = 0; i < byteCount; i++) {
    let b = bf.bits[i];
    while (b !== 0) {
      b = b & (b - 1);
      c++;
    }
  }
  const padding = byteCount * 8 - bf.m;
  if (padding > 0) {
    const lastByte = bf.bits[byteCount - 1];
    const validMask = (1 << (8 - padding)) - 1;
    let paddingBits = lastByte & ~validMask;
    while (paddingBits !== 0) {
      paddingBits = paddingBits & (paddingBits - 1);
      c--;
    }
  }
  return c;
}

function saturation(bf) {
  return popcount(bf) / bf.m;
}

function falsePositiveRate(bf) {
  if (bf.n === 0) return 0.0;
  return Math.pow(1.0 - Math.exp((-bf.k * bf.n) / bf.m), bf.k);
}

function isEmpty(bf) { return bf.n === 0; }

function union(bf1, bf2) {
  if (bf1.m !== bf2.m || bf1.k !== bf2.k) {
    throw new Error('BloomFilter.union: incompatible filters');
  }
  const byteCount = Math.ceil(bf1.m / 8);
  const newBits = new Uint8Array(byteCount);
  for (let i = 0; i < byteCount; i++) {
    newBits[i] = bf1.bits[i] | bf2.bits[i];
  }
  return Object.freeze({ bits: newBits, m: bf1.m, k: bf1.k, n: bf1.n + bf2.n });
}

function ofList(list, opts = {}) {
  return list.reduce((bf, x) => add(x, bf), create(opts));
}

function memAll(list, bf) {
  return list.every(x => mem(x, bf));
}

function memAny(list, bf) {
  return list.some(x => mem(x, bf));
}

function clear(bf) {
  return create({ m: bf.m, k: bf.k });
}

function copy(bf) {
  return Object.freeze({
    bits: new Uint8Array(bf.bits),
    m: bf.m,
    k: bf.k,
    n: bf.n,
  });
}

function toString(bf) {
  return `BloomFilter(m=${bf.m}, k=${bf.k}, n=${bf.n}, saturation=${(saturation(bf) * 100).toFixed(2)}%)`;
}

module.exports = {
  create, createOptimal, add, mem, count, size, numHashes,
  popcount, saturation, falsePositiveRate, isEmpty, union,
  ofList, memAll, memAny, clear, copy, toString,
  _hash1: hash1, _hash2: hash2, _hashI: hashI,
  _getBit: getBit, _setBit: setBit,
};
